### DISCUSSION TITLE
Doing a bunch of async things in any order? Like "concurrent" in Hack, but in C#?

### DISCUSSION THREAD

**Discussion by bherila** [dotnet/csharplang#4009] [Q&A] [answered]
In Hacklang, I can do something like 

```hack
concurrent{
  await thing1(..);
  await thing2(..);
  await thing3(..);
  await async {
    // thing4 async code (lambda)
  };
}
```

Ref: https://docs.hhvm.com/hack/asynchronous-operations/concurrent

These are executed in any order and the block continues after all the `await`s are done. Of course there must be no dependencies between the items in the concurrent block. But in many cases this is just fine.

Is there a clean way to do this in C#?

Thanks!

**Comment by HaloFour** [comment]
```cs
async Task Foo() {
    var task1 = Thing1();
    var task2 = Thing2();
    var task3 = Thing3();

    return Task.WhenAll(task1, task2, task3);
}
```

**Comment by CyrusNajmabadi** [comment, marked as answer]
```c#
await Task.WhenAll(thing1(), thing2(), thing3());
```

**Reply by bherila** [reply]
Did not notice that overload on Task.WhenAll. 💯 
P.S. Hello!

**Reply by CyrusNajmabadi** [reply]
Hi!  I didn't even see it was you Ben :)   Hope you're doing well :)

**Comment by theunrepentantgeek** [comment]
With [suitable extension methods](https://gist.github.com/jnm2/3660db29457d391a34151f764bfe6ef7) on `ValueTuple`, you can write this:

```
var (t1, t2, t3, t4) = await (thing1Async(), thing2Async(), thing3Async, thing4Async)
```

**Comment by bherila** [comment]
Thank you to everyone for the super fast responses!
